abe/ma-abe.go

```diff
@@ -0,0 +1,492 @@
+// attributes can be spread across many different authorities, eliminating the
+// need for a central authority. Secret keys, each connected to a single
+// attribute, are generated by the relevant authorities, such that only a set
+// of keys whose attributes are sufficient according to the boolean formula can
+// decrypt the message.
+
+// MAABE represents a MAABE scheme.
+type MAABE struct {
+    P *big.Int
+    g1 *bn256.G1
+    g2 *bn256.G2
+}
+
+// NewMAABE configures a new instance of the scheme.
+func NewMAABE() *MAABE {
+    return &MAABE{
+            P: bn256.Order,
+            g1: new(bn256.G1).ScalarBaseMult(big.NewInt(1)),
+            g2: new(bn256.G2).ScalarBaseMult(big.NewInt(1)),
+    }
+}
+
+// MAABEPubKey represents a public key for an authority.
+type MAABEPubKey struct {
+    Attribs []string
+    EggToAlpha map[string]*bn256.GT
+    GToY map[string]*bn256.G2
+}
+
+// MAABESecKey represents a secret key for an authority.
+type MAABESecKey struct {
+    Attribs []string
+    Alpha map[string]*big.Int
+    Y map[string]*big.Int
+}
+
+// MAABEAuth represents an authority in the MAABE scheme.
+type MAABEAuth struct {
+    Id string
+    Pk *MAABEPubKey
+    Sk *MAABESecKey
+}
+
+// NewMAABEAuth configures a new instance of an authority and generates its
+// public and secret keys for the given set of attributes. In case of a failed
+// procedure an error is returned.
+func (a *MAABE) NewMAABEAuth(id string, attribs []string) (*MAABEAuth, error) {
+    numattrib := len(attribs)
+    // sanity checks
+    if numattrib == 0 {
+        return nil, fmt.Errorf("empty set of authority attributes")
+    }
+    if len(id) == 0 {
+        return nil, fmt.Errorf("empty id string")
+    }
+    // rand generator
+    sampler := sample.NewUniform(a.P)
+    // generate seckey
+    alpha_i, err := data.NewRandomVector(numattrib, sampler)
+    if err != nil {
+        return nil, err
+    }
+    y_i, err := data.NewRandomVector(numattrib, sampler)
+    if err != nil {
+        return nil, err
+    }
+    alpha := make(map[string]*big.Int)
+    y := make(map[string]*big.Int)
+    for i, at := range attribs {
+        alpha[at] = alpha_i[i]
+        y[at] = y_i[i]
+    }
+    // generate pubkey
+    eggToAlpha := make(map[string]*bn256.GT)
+    gToY := make(map[string]*bn256.G2)
+    for _, at := range attribs {
+        eggToAlpha[at] = new(bn256.GT).ScalarMult(bn256.Pair(a.g1, a.g2), alpha[at])
+        gToY[at] = new(bn256.G2).ScalarMult(a.g2, y[at])
+    }
+    sk := &MAABESecKey{Attribs: attribs, Alpha: alpha, Y: y}
+    pk := &MAABEPubKey{Attribs: attribs, EggToAlpha: eggToAlpha, GToY: gToY}
+    return &MAABEAuth{
+        Id: id,
+        Pk: pk,
+        Sk: sk,
+    }, nil
+}
+
+// AddAttribute generates public and secret keys for a new attribute that is
+// given as input. In case of a failed procedure an error is returned, and nil
+// otherwise.
+func (auth *MAABEAuth) AddAttribute(attrib string, maabe *MAABE) error {
```

Comment:
I suggest to put `*MAABE` into `MAABEAuth` so that it doesn't need to be carried around.